### PR TITLE
upgrade(depot component) use typehead in depot component

### PR DESCRIPTION
**Pull request by jeremielodi**
Optimizes depot component when there is a lot of depots registered

### CODE REVIEW

**Comment by jeremielodi** [comment]
do not merge this PR yet

File: client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js
Comment (jniles) [inline client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js]: This should be optional `<?`.

File: client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js
Comment (jniles) [inline client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js]: Just a couple of notes:
 1. When using angular, you need to make sure that angular is aware of changes you are making.  Remember that angular needs to recompute things in a $digest loop.
a. Instead of `setTimeout` use [`$timeout()`](https://docs.angularjs.org/api/ng/service/$timeout).  
b. Instead of `document`, use [`$document`](https://docs.angularjs.org/api/ng/service/$document).  This makes it easier to test components in unit tests.

 2. This `.keyup()` even is never removed.  If you set up an event in the `$onInit()`, you need to remove it in [`$onDestroy()`](https://docs.angularjs.org/guide/component#component-based-application-architecture).

File: client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js
Comment (jniles) [inline client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js:33]: I think there is a typo here and this should be `depotsSelected`.

File: client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js
Comment (jniles) [reply client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js]: However, you actually don't need any of this code - uib-typeahead will do all this for you.  All you need to do is pass a function that returns an `$http` request to the `uib-typeahead` directive.  We do this is in [bhFindPatient](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/modules/templates/bhFindPatient.tmpl.html#L46).

It looks like this:
```html
 uib-typeahead="patient as patient.display_name for patient in $ctrl.searchByName($viewValue)"
```
where `$ctrl.searchByName` is [this function](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/js/components/bhFindPatient.js#L165):
```js
  function searchByName(text) {
    $ctrl.loadStatus = 'loading';

    // format query string parameters
    const options = {
      display_name : text.toLowerCase(),
      limit        : LIMIT,
    };

    return Patients.searchByName(options);
};
```

This means you can just set `loadDepots()` as the function to call with `$viewValue` and it will then load the required depots.

File: client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js
Comment (jniles) [inline client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js]: It doesn't look like you need this eslint-disable directive.

File: server/config/routes.js
Comment (jniles) [inline server/config/routes.js]: Why write an entirely new route?  Can't you just add the `uuids` parameter to the filters in the [/depots](https://github.com/IMA-WorldHealth/bhima/blob/master/server/controllers/inventory/depots/index.js#L113) route?

Otherwise, you'll need to write a couple new integration tests for this new route.

File: client/src/js/components/bhDepotSelectSearch/bhDepotSelectSearch.html
Comment (jniles) [inline client/src/js/components/bhDepotSelectSearch/bhDepotSelectSearch.html]: Looks like a typo here.  I think this should be `bh-multiple-depot-search-select` not `seach`

File: test/end-to-end/shared/components/bhMultipleDepotSearchSelect.js
Comment (jniles) [inline test/end-to-end/shared/components/bhMultipleDepotSearchSelect.js]: Both of these `input` operations are asynchronous and need `await` before them.

File: server/controllers/inventory/depots/index.js
Comment (jniles) [inline server/controllers/inventory/depots/index.js]: Remember that passing in raw values always exposes you to SQL injection attacks.  You should always escape values before putting them into a string to give to the database.

File: server/controllers/inventory/depots/index.js
Comment (jniles) [inline server/controllers/inventory/depots/index.js]: You can do this entire thing by added only two lines to the `list()` function.

The first is changing
```js
  const options = db.convert(req.query, ['uuid']);
```
to 
```js
  const options = db.convert(req.query, ['uuid', 'uuids']);
```

The second is adding a line below:
```js
  filters.equals('uuid', 'uuid', 'd');
```
add this line:
```js
  filters.equals('uuid', 'uuid', 'd');
  filters.equals('uuids', 'uuid', 'd', true);
```
Everything should now work as expected.

**Review by jniles — CHANGES_REQUESTED** [review]
@jeremielodi, thanks for working on this.

I've made a few comments.  The two big issues to address here are:
  1. Using the uib-typeaheads functionality to send the HTTP request, rather than using setTimeout + keyup().  It should function the same way as bhFindPatient.
  2. Using the `/depots` route instead of creating your own.  It should result in only two extra lines of code.

Let me know when you need another review.

File: client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js
Comment (jeremielodi) [reply client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js]: thanks

File: server/controllers/admin/users/depots.js
Comment (jniles) [inline server/controllers/admin/users/depots.js:52]: Good fix 👍

File: client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js
Comment (jniles) [inline client/src/js/components/bhDepotSelectSearch/bhDepotSelect.js:8]: Is required optional?  If so, it should be labeled that way `<?`.

File: server/controllers/inventory/depots/index.js
Comment (jniles) [inline server/controllers/inventory/depots/index.js:159]: If you don't have this function, what happens?  If you just put in 
```js
filters.equals('uuids', 'uuid', 'd', true)
```
doesn't that work?  I thought it should.

File: server/lib/filter.js
Comment (jniles) [inline server/lib/filter.js:148]: 👍

File: test/end-to-end/staffingIndice/payroll-indice.spec.js
Comment (jniles) [inline test/end-to-end/staffingIndice/payroll-indice.spec.js:29]: 👍

**Review by jniles — COMMENTED** [review]
Just one quick check on equals and this looks good.  I notice we don't have unit tests for the `filters.equals()` case, so I wouldn't be surprised if the functionality stopped working silently at some point.

File: server/controllers/inventory/depots/index.js
Comment (jeremielodi) [reply server/controllers/inventory/depots/index.js:159]: An array can have several items but our filter class doesn't add the corresponding number of question marks <b>?</b>


Maybe we should update ```filters.equals```

Comment (jniles) [reply server/controllers/inventory/depots/index.js:159]: It should add an SQL `IN ()` clause instead of an equals.  Isn't that what you need?  [Here is the code](https://github.com/IMA-WorldHealth/bhima/blob/8e8e8bf541ab3408a51cbec06ea8de7c379a68ea/server/lib/filter.js#L147).

**Comment by mbayopanda** [comment]
- Is it possible to have the dropdown of depots list over the footer bar

image

**Comment by mbayopanda** [comment]
It is very fast now 👍

File: server/controllers/inventory/depots/index.js
Comment (jeremielodi) [reply server/controllers/inventory/depots/index.js:159]: In can see very well that the number of ```  '?'``` doesn't change if we have an array.
We can comeback to this issue in another PR I think.

**Comment by jeremielodi** [comment]
Let me try to use append-to-body if it will work in this case

**Comment by jeremielodi** [comment]
good-sele

**Review by jniles — APPROVED** [review]

**Review by mbayopanda — APPROVED** [review]
bors r+

**Pull request merged**
